upload-api/billing.js

```diff
@@ -0,0 +1,89 @@
+  /**
+   * @param {string} [message] Context for the message.
+   * @param {ErrorOptions} [options]
+   */
+  constructor(message, options) {
+    super(undefined, options)
+    this.name = /** @type {const} */ ('InvalidSubscriptionState')
+    this.detail = message
+  }
+
+  describe() {
+    return `subscription cannot be updated because it is not in a valid state: ${this.detail}`
+  }
+}
+
+export class BillingProviderUpdateError extends Failure {
+  /**
+   * @param {string} [message] Context for the message.
+   * @param {ErrorOptions} [options]
+   */
+  constructor(message, options) {
+    super(undefined, options)
+    this.name = /** @type {const} */ ('BillingProviderUpdateError')
+    this.detail = message
+  }
+
+  describe() {
+    return `encountered an error updating subscription: ${this.detail}`
+  }
+}
+
+/**
+ * 
+ * @param {import('stripe').Stripe} stripe 
+ * @returns {import("./types").BillingProvider}
+ */
+export function createStripeBillingProvider(stripe) {
+  return {
+    async hasCustomer(customer) {
+      const customersResponse = await stripe.customers.list({ email: toEmail(/** @type {import('@web3-storage/did-mailto').DidMailto} */(customer)) })
+      return { ok: (customersResponse.data.length > 0) }
+    },
+
+    async setPlan(customerDID, plan) {
+      /** @type {import('stripe').Stripe.SubscriptionItem[] | undefined} */
+      let subscriptionItems
+      /** @type {string | undefined} */
+      let priceID
+      try {
+        const prices = await stripe.prices.list({ lookup_keys: [plan] })
+        priceID = prices.data.find(price => price.lookup_key === plan)?.id
+        if (!priceID) return (
+          { error: new InvalidSubscriptionState(`could not find Stripe price with lookup_key ${plan} - cannot set plan`) }
+        )
+
+        const email = toEmail(/** @type {import('@web3-storage/did-mailto').DidMailto} */(customerDID))
+        const customers = await stripe.customers.list({ email, expand: ['data.subscriptions'] })
+        if (customers.data.length != 1) return (
+          { error: new InvalidSubscriptionState(`found ${customers.data.length} Stripe customer(s) with email ${email} - cannot set plan`) }
+        )
+
+        const customer = customers.data[0]
+        const subscriptions = customer.subscriptions?.data
+        if (subscriptions?.length != 1) return (
+          { error: new InvalidSubscriptionState(`found ${subscriptions?.length} Stripe subscriptions(s) for customer with email ${email} - cannot set plan`) }
+        )
+
+        subscriptionItems = customer.subscriptions?.data[0].items.data
+        if (subscriptionItems?.length != 1) return (
+          { error: new InvalidSubscriptionState(`found ${subscriptionItems?.length} Stripe subscriptions item(s) for customer with email ${email} - cannot set plan`) }
+        )
+      } catch (/** @type {any} */ err) {
+        return { error: new InvalidSubscriptionState(err.message, { cause: err }) }
```

Comment:
It seems we've handled all expected invalid states - if any of these calls to stripe throw it's because of a network or config error or something. I would just not catch this and let it be thrown (and caught and handled by ucanto).

Reply:
Oh I didn't know Ucanto had default handling for errors here, to be honest. A couple thoughts:

1) I'm not sure this `BillingProvider` will always be used in Ucanto - I'm using the `Result` type here because @Gozala has been advocating for it but otherwise this is something I'd like to use, eg, in the Stripe webhook where top level error handling will be different
2) I've found it's annoying when trying to call a function that might either return an `{error: ...}` response OR throw an exception - I think @Gozala has also been advocating for trying to make sure all errors in functions that return `Result`s are caught and turned into typed `{error: ...}` style responses. I could totally be going overboard here! Curious for your thoughts @Gozala!

Reply:
We talked about this async and came up with this: https://github.com/web3-storage/RFC/pull/7